File: src/main/java/com/leet/hard/WordBreakII.java
package com.leet.hard;

/**
 * 140. Word Break II Hard
 *
 * https://leetcode.com/problems/word-break-ii/
 *
 * Given a string s and a dictionary of strings wordDict, add spaces in s to
 * construct a sentence where each word is a valid dictionary word. Return all
 * such possible sentences in any order.
 *
 * Note that the same word in the dictionary may be reused multiple times in the
 * segmentation.
 *
 *
 *
 * Example 1:
 *
 * Input: s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"] Output:
 * ["cats and dog","cat sand dog"]
 *
 * Example 2:
 *
 * Input: s = "pineapplepenapple", wordDict =
 * ["apple","pen","applepen","pine","pineapple"] Output: ["pine apple pen
 * apple","pineapple pen apple","pine applepen apple"] Explanation: Note that
 * you are allowed to reuse a dictionary word.
 *
 * Example 3:
 *
 * Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"] Output:
 * []
 *
 *
 *
 * Constraints:
 *
 * 1 <= s.length <= 20 1 <= wordDict.length <= 1000 1 <= wordDict[i].length <=
 * 10 s and wordDict[i] consist of only lowercase English letters. All the
 * strings of wordDict are unique.
 *
 *
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

public class WordBreakII {
	Map<Integer, List<String>> dp = new HashMap<>();

	public List<String> solve(String s, int position, HashSet<String> dict) {
		if (position >= s.length()) {
			return new ArrayList<>();
		}

		if (dp.containsKey(position)) {
			return dp.get(position);
		}

		// form the next word
		String collect = "";
		List<String> ans = new ArrayList<>();
		for (int i = position; i < s.length(); i++) {
			collect += s.charAt(i);
			if (dict.contains(collect)) {
				if (i == s.length() - 1) {
					ans.add(collect.toString());
				} else {
					List<String> cur = solve(s, i + 1, dict);
					if (cur.size() != 0) {
						for (String word : cur) {
							ans.add(collect + " " + word);
						}
					}
				}
			}
		}

		dp.put(position, ans);
		return ans;
	}

	public List<String> wordBreak(String s, List<String> wordDict) {
		var dict = new HashSet<String>(wordDict);
		var ans = solve(s, 0, dict);
		System.out.println(dp);
		return ans;
	}

	public static void main(String[] args) {
		var ans = new WordBreakII().wordBreak("pineapplepenapple",
				List.of("apple", "pen", "applepen", "pine", "pineapple"));
		System.out.println(ans);
	}
}

File: src/main/java/com/leet/hard/WordBreakII.ts
/** A node in the Trie */
class TrieNode {
	leaf: boolean = false;
	nodes: { [val: string]: TrieNode } = {};
}

/** Trie data structure */
class Trie {
	root: TrieNode;
	dp: { [pos: number]: number[] };

	constructor() {
		this.root = new TrieNode();
		this.dp = {};
	}

	/** Add a single word to trie */
	public addWord(word: string) {
		let node = this.root;
		for (const ch of word) {
			if (!node.nodes[ch]) {
				node.nodes[ch] = new TrieNode();
			}
			node = node.nodes[ch];
		}
		node.leaf = true;
	}

	/** Add a bunch of words */
	addAllWords(words: string[]) {
		words.forEach((word) => this.addWord(word));
	}

	/** We start matching the sentence at a given position */
	private matchAtPosition(sentence: string, pos: number): number[] {
		if (this.dp[pos]) {
			return this.dp[pos];
		}

		const ans: number[] = [];

		/** If already beyond sentence size, we abort */
		if (pos > sentence.length) {
			return ans;
		}

		let node = this.root;
		for (let i = pos; i < sentence.length; i++) {
			const ch = sentence[i];
			node = node.nodes[ch];
			if (!node) {
				break;
			}
			/** If current node is leaf, we can mark this position as end of a sentence */
			if (node?.leaf) {
				ans.push(i);
			}
		}

		return ans;
	}

	/**
	 * Match a sentence
	 * - We save all answers in ans
	 * - wordBreaks hold all the breaks till this point (points in the sentence where a word exists)
	 * - pos is the current position in the sentence
	 * - size is the active size of the wordBreaks array (optimisation hack - we use same array for all iterations)
	 **/
	match(sentence: string, ans: string[] = [], wordBreaks: number[] = [], pos = 0, size = 1): string[] {
		/** if a complete match, we break the sentence into words and join them with spaces */
		if (pos === sentence.length) {
			const collect = [];
			for (let i = 1; i < size; i++) {
				let cur = wordBreaks[i], prev = wordBreaks[i - 1];
				collect.push(sentence.slice(prev, cur));
			}
			ans.push(collect.join(' '));
			return ans;
		}

		/** Optimisation: As we use the same wordBreaks array, it may be a fresh one or an old one */
		if (wordBreaks.length < size) {
			wordBreaks.push(pos); // inserts at `size` index, a placeholder value
		}

		/** Get all matching word breaks for this position */
		const matches = this.matchAtPosition(sentence, pos);
		for (const match of matches) {
			// replace the placeholder value
			wordBreaks[size] = match + 1;
			// add this as break point, and match the sentence onwards
			this.match(sentence, ans, wordBreaks, match + 1, size + 1);
		}

		return ans;
	}
}

function wordBreak(s: string, wordDict: string[]): string[] {
	const trie = new Trie();
	trie.addAllWords(wordDict);
	return trie.match(s);
};

function main() {
	[
		{ s: "pineapplepenapple", d: ["apple", "pen", "applepen", "pine", "pineapple"] },
		{ s: "catsandog", d: ["cats", "dog", "sand", "and", "cat"] }
	].forEach(({ s: sentence, d: wordDict }) => {
		console.log({ sentence, wordDict, ans: wordBreak(sentence, wordDict) });
	});
}

main()